### PR TITLE
Fix pytest.raises handling of unicode exceptions in Python 2

### LINKED ISSUE #5478: cannot make unicode match assertions with pytest.raises python2 pytest
```
    def test_u():
        with pytest.raises(AssertionError, match=u"\u2603"):
>           assert False, u"\u2603"
E           UnicodeEncodeError: 'ascii' codec can't encode character u'\u2603' in position 0: ordinal not in range(128)
```

### PR DESCRIPTION
**Pull request by graingert**
Fixes #5478

### CODE REVIEW

**Comment by graingert** [comment]
I think I've missed the boat on Py2 fixes...

**Comment by blueyed** [comment]
You could create a PR against the 4.6-maintenace branch I guess.
But it should really have a test then - and/or fix the failing ones (have not looked at CI).

**Comment by graingert** [comment]
I think this is against the 4.6 maintenance branch

**Comment by nicoddemus** [comment]
Only linting is failing. To fix it, either:

1. Run `tox -e linting` and commit the changes.
2. Install `pre-commit` and run `pre-commit install` so each commit gets linted automatically.

File: src/_pytest/_code/code.py [hunk 1]
```diff
@@ -572,6 +572,11 @@ def match(self, regexp):
         raised.
         """
         __tracebackhide__ = True
+        if isinstance(regexp, text_type):              
+            if not re.search(regexp, text_type(self.value)):
+                assert 0, u"Pattern '{!s}' not found in '{!s}'".format(regexp, self.value)
+            return True
+
         if not re.search(regexp, str(self.value)):
```
Comment (nicoddemus) [inline src/_pytest/_code/code.py]: Actually I think the only change required is to use `text_type(self.value)` instead of `str(self.value)` here.

Comment (graingert) [reply src/_pytest/_code/code.py]: This doesn't work when you have a non-ascii bytes exception message

Comment (nicoddemus) [reply src/_pytest/_code/code.py]: Oh you are absolutely right, `self.value` might be already encoded. So to avoid duplication I recommend to change to this:

```python
        __tracebackhide__ = True
        value = text_type(self.value) if isinstance(regexp, text_type) else str(self.value)
        if not re.search(regexp, value):
            assert 0, "Pattern '{!s}' not found in '{!s}'".format(regexp, value)
        return True
```

Comment (graingert) [reply src/_pytest/_code/code.py]: But then the formatting won't be safe

Comment (nicoddemus) [reply src/_pytest/_code/code.py]: You are right, but I think this solves it:

```python
assert 0, u"Pattern '{!s}' not found in '{!s}'".format(regexp, value)
```

It should be safe to always use unicode messages with the assert statement

Comment (pombredanne) [reply src/_pytest/_code/code.py]: or you could use a repr instead?
`assert 0, u"Pattern {!r} not found in {!r}".format(regexp, value)`

Comment (RonnyPfannschmidt) [reply src/_pytest/_code/code.py]: how about safe_str ? or one of the to ascii helper s we have in the python2 support code?

**Comment by graingert** [comment]
safe_str looks best

**Comment by graingert** [comment]
```
ERROR: FAIL could not package project - v = InvocationError(u"/home/travis/build/pytest-dev/pytest/.tox/.package/bin/python -m pip install setuptools 'setuptools>=40.0'", 1)
```
eh?

**Comment by blueyed** [comment]
py36-xdist:
```
_________________________ TestRaises.test_raises_match _________________________
[gw0] linux -- Python 3.6.7 /home/travis/build/pytest-dev/pytest/.tox/py36-xdist/bin/python
self = <raises.TestRaises object at 0x7f9829898d68>
    def test_raises_match(self):
        msg = r"with base \d+"
        with pytest.raises(ValueError, match=msg):
            int("asdf")
    
        msg = "with base 10"
        with pytest.raises(ValueError, match=msg):
            int("asdf")
    
        msg = "with base 16"
        expr = r"Pattern '{}' not found in 'invalid literal for int\(\) with base 10: 'asdf''".format(
            msg
        )
        with pytest.raises(AssertionError, match=expr):
            with pytest.raises(ValueError, match=msg):
>               int("asdf", base=10)
E               AssertionError: Pattern "Pattern 'with base 16' not found in 'invalid literal for int\\(\\) with base 10: 'asdf''" not found in 'Pattern \'with base 16\' not found in "invalid literal for int() with base 10: \'asdf\'"'
```

**Comment by nicoddemus** [comment]
@graingert the previous failure was due to a bug in tox 3.13.0. 3.13.1 has just been out with a fix, so please try again!

**Comment by graingert** [comment]
@nicoddemus is this fix still desired?

**Comment by nicoddemus** [comment]
I don't mind it. 😁 

@asottile are you against this? I understand some parts of the standard library might break with this, but I'm 👍 because we can do better than unittest and with a small change to boot.

**Comment by asottile** [comment]
I'm conflicted, because exceptions with unicode messages in python 2 are wrong, but this enables something that seems like it should work. so -0 from me, feel free to go ahead

File: testing/python/raises.py [hunk 2]
```diff
@@ -278,3 +278,7 @@ def __class__(self):
                 with pytest.raises(CrappyClass()):
                     pass
             assert "via __class__" in excinfo.value.args[0]
+
+    def test_u(self):
```
Comment (graingert) [inline testing/python/raises.py]:
```
______________________________ TestRaises.test_u ______________________________

self = <raises.TestRaises object at 0x0000000007B297B8>

    def test_u(self):
        with pytest.raises(AssertionError, match=u"\u2603"):
>           assert False, u"\u2603"
E           UnicodeEncodeError: 'ascii' codec can't encode character u'\u2603' in position 0: ordinal not in range(128)
```

seems to be failing on windows CI

Comment (nicoddemus) [reply testing/python/raises.py]: I've applied this patch:

```diff
diff --git a/src/_pytest/_code/code.py b/src/_pytest/_code/code.py
index e621f3ee0..407155373 100644
--- a/src/_pytest/_code/code.py
+++ b/src/_pytest/_code/code.py
@@ -574,7 +574,7 @@ class ExceptionInfo(object):
         __tracebackhide__ = True
         value = safe_str(self.value)
         if not re.search(regexp, value):
-            assert 0, "Pattern '{!s}' not found in '{!s}'".format(regexp, value)
+            assert 0, u"Pattern {!r} not found in {!r}".format(regexp, value)
         return True
```

The problem is that even so the check still fails, when it should pass:

```
______________________________ test_u _______________________________

    def test_u():
        with pytest.raises(AssertionError, match=u"\u2603"):
>           assert False, u"\u2603"
E           AssertionError: Pattern u'\u2603' not found in '\xe2\x98\x83\nassert False'

.tmp\test_u.py:6: AssertionError
```

This will be true for any system which can't handle that unicode string using the default locale.

This solution works for me and the rest of the suite:

```
```

File: changelog/5478.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Use safe_str to serialize Exceptions in pytest.raises
```
Comment (nicoddemus) [inline changelog/5478.bugfix.rst]:
```suggestion
Fix encode error when using unicode strings in exceptions with ``pytest.raises``.
```

File: testing/python/raises.py (hunk 2 above)
Comment (asottile) [reply testing/python/raises.py]: though this probably triggers the same error I posted in the issue when using `--tb=native` since now we're raising a (n incorrect) unicode message in python 2 🤔

File: src/_pytest/_code/code.py [hunk 4]
```diff
@@ -572,8 +572,9 @@ def match(self, regexp):
         raised.
         """
         __tracebackhide__ = True
-        if not re.search(regexp, str(self.value)):
-            assert 0, "Pattern '{!s}' not found in '{!s}'".format(regexp, self.value)
+        value = safe_str(self.value)
+        if not re.search(regexp, value):
+            assert 0, "Pattern '{!s}' not found in '{!s}'".format(regexp, value)
```
Comment (RonnyPfannschmidt) [inline src/_pytest/_code/code.py]: option or followup: we might want to use `raise AssertionError(...)` in that case

**Comment by graingert** [comment]
@nicoddemus I don't really have the time to put into this pr right now, could you take it over?

File: testing/python/raises.py (hunk 2 above)
Comment (nicoddemus) [reply testing/python/raises.py]: Thanks, but actually it works:

```
 λ pytest .tmp\test_u.py  -q
.
1 passed in 0.02 seconds
```

```
λ pytest .tmp\test_u.py  -q --tb=native
.
1 passed in 0.02 seconds
```


 so I think we can go with that solution. 👍

File: src/_pytest/_code/code.py (hunk 4 above)
Comment (nicoddemus) [reply src/_pytest/_code/code.py]: Done

Out of curiosity, is there any reason other than style to prefer one over the other?

File: testing/python/raises.py [hunk 5]
```diff
@@ -278,3 +278,9 @@ def __class__(self):
                 with pytest.raises(CrappyClass()):
                     pass
             assert "via __class__" in excinfo.value.args[0]
+
+    def test_unicode_message(self):
+        """pytest.raises should be able to match unicode messages when using a unicode regex (#5478)
+        """
+        with pytest.raises(AssertionError, match=u"\u2603"):
+            assert False, u"\u2603"
```
Comment (asottile) [inline testing/python/raises.py]: can you also test a failure of unicode<=>unicode as well?

Comment (nicoddemus) [reply testing/python/raises.py]: Sure, but what do you mean by `unicode<=>unicode`? 🤔

Comment (asottile) [reply testing/python/raises.py]: oh yeah that is super unclear, what the heck did I mean 🤔 

there's 6(?) combinations I think and maybe need more tests for them:

- bytes exception bytes regex success
- bytes exception bytes regex failure
- unicode exception unicode regex success
- unicode exception unicode regex failure (I suspect that this currently causes an undisplayable exception in python2 because *pytest* would be raising a unicode exception)
- ascii unicode exception ascii unicode bytes success (only in py2 due to implicit py2 conversion)
- ascii unicode exception ascii unicode bytes failure (only in py2 due to implicit py2 conversion)

Comment (nicoddemus) [reply testing/python/raises.py]: I see, thanks. Done.

Surprisingly on Python 3 we don't support raising exceptions using bytes (`raise RuntimeError(b"foo")`), which I find reasonable actually.

File: src/_pytest/_code/code.py [hunk 6]
```diff
@@ -572,8 +572,13 @@ def match(self, regexp):
...
         """
         __tracebackhide__ = True
-        if not re.search(regexp, str(self.value)):
-            assert 0, "Pattern '{!s}' not found in '{!s}'".format(regexp, self.value)
+        value = (
+            text_type(self.value) if isinstance(regexp, text_type) else str(self.value)
+        )
+        if not re.search(regexp, value):
+            raise AssertionError(
+                u"Pattern '{}' not found in '{}'".format(regexp, value)
```
Comment (asottile) [inline src/_pytest/_code/code.py]: In python2 this will raise on non-ascii bytes

Comment (nicoddemus) [reply src/_pytest/_code/code.py]: Hmm indeed. Any suggestions?

I guess we can use `safe_str` to display the messages? Although `safe_str` garbles on invalid output, which might end up being more confusing.

Alternatively we can try to decode it, and upon failure, just show `<decode error>` instead of the original string:

```
Pattern 'hello' not found in <<decode error>>
```

Or something like this.

Comment (asottile) [reply src/_pytest/_code/code.py]: !r might be the best bet :S

Comment (nicoddemus) [reply src/_pytest/_code/code.py]: Done using `{!r}` for formatting instead, which is the safest approach I think.

Comment (graingert) [reply src/_pytest/_code/code.py]: > Done using `{!r}` for formatting instead, which is the safest approach I think.

I think that change will need to be ported to pytest 5

Comment (asottile) [reply src/_pytest/_code/code.py]: probably should, yeah -- there's currently this weird output due to not using `!r`

```
E           AssertionError: Pattern ''foo' not found in ''bar'
```

```python
import pytest


def test():
    with pytest.raises(AssertionError, match="'foo"):
        raise AssertionError("'bar")
```

Comment (nicoddemus) [reply src/_pytest/_code/code.py]: I don't think this should be ported to pytest 5... this is a non-issue on Python 3, I believe.

Comment (asottile) [reply src/_pytest/_code/code.py]: I created #5553
